### PR TITLE
Replace as_area_btree with ordered dictionary.

### PR DESCRIPTION
**Pull request by jxsvoboda**
Anybody else interested &&/|| willing to review this change just let me know.

### CODE REVIEW

File: kernel/generic/src/mm/as.c [hunk 1]
```diff
@@ -313,76 +377,39 @@ NO_TRACE static bool check_area_conflicts(as_t *as, uintptr_t addr,
...
+
+		/* Next area */
+		odlink = odict_next(odlink, &as->as_areas);
 	}
 
-	node = btree_leaf_node_right_neighbour(&as->as_area_btree, leaf);
-	if (node) {
-		area = (as_area_t *) node->value[0];
+	/* Next area, if any, is the first with base > than our base address */
+	if (odlink != NULL) {
```
Comment (jermar) [inline kernel/generic/src/mm/as.c]: How does hit work if there was no area with base <= our base address, but there is an area with base > our base address? It seems to me odlink will be NULL and so the branch for the latter will not get executed.

File: kernel/generic/src/mm/as.c [hunk 2]
```diff
@@ -487,33 +482,30 @@ NO_TRACE static uintptr_t as_get_unmapped_area(as_t *as, uintptr_t bound,
...
-	list_foreach(as->as_area_btree.leaf_list, leaf_link, btree_node_t, node) {
-
-		for (btree_key_t i = 0; i < node->keys; i++) {
-			as_area_t *area = (as_area_t *) node->value[i];
+	as_area_t *area = as_area_first(as);
+	while (area != NULL) {
+		mutex_lock(&area->lock);
 
-			mutex_lock(&area->lock);
+		addr = ALIGN_UP(area->base + P2SZ(area->pages), PAGE_SIZE);
```
Comment (jermar) [inline kernel/generic/src/mm/as.c]: area->base + P2SZ(area->pages) is already aligned.

File: kernel/generic/src/mm/as.c (hunk 1 above)
Comment (jxsvoboda) [reply kernel/generic/src/mm/as.c]: Nice catch, fixed.

File: kernel/generic/src/mm/as.c (hunk 2 above)
Comment (jxsvoboda) [reply kernel/generic/src/mm/as.c]: I guess if you trust area->base to be aligned, then why not.

**Pull request merged**